File: src/main/java/com/raychenon/leetcode/ValidParentheses.java
package com.raychenon.leetcode;

import java.util.*;

/**
 * User: raychenon
 * Date: 2019-09-04
 * https://leetcode.com/problems/valid-parentheses/
 */
public class ValidParentheses {

    /**
     * time complexity : O(n)
     * space complexity : O(n) for only opening parentheses
     * The trick is to store the different parentheses in FIFO structure: Stack
     * An opening parenthese must be closed by its closing counterpart.
     */
    public boolean isValid(String s) {
        if (s == null || s.isEmpty()) return true;

        Map<Character, Character> map = new HashMap<Character, Character>();
        map.put('}', '{');
        map.put(')', '(');
        map.put(']', '[');
        map.put('>', '<');

        Set<Character> openingSet = new HashSet<Character>(Arrays.asList('(', '{', '[', '<'));

        Stack<Character> stack = new Stack<Character>();
        for (Character c : s.toCharArray()) {
            if (map.containsKey(c)) {
                if (!stack.isEmpty()) {
                    if (map.get(c) == stack.peek()) {
                        stack.pop();
                    }
                } else {
                    // if more closing parenthese, can already fast-exit
                    return false;
                }
            }
            if (openingSet.contains(c)) {
                stack.push(c);
            }
        }
        // the amount of closing nullify the count of opening parentheses , stack should be empty
        return stack.isEmpty();
    }
}

File: src/main/kotlin/com/raychenon/kotlin/leetcode/ValidParentheses.kt
package com.raychenon.kotlin.leetcode

import java.util.Stack

/**
 * User: raychenon
 * Date: 2019-09-05
 * https://leetcode.com/problems/valid-parentheses/
 */
object ValidParentheses {

    /**
     * time complexity : O(n)
     * space complexity : O(n) for only opening parentheses
     * The trick is to store the different parentheses in FIFO structure: Stack
     * An opening parenthese must be closed by its closing counterpart.
     */
    fun isValid(s: String): Boolean {
        if (s.isEmpty()) return true

        val stack = Stack<Char>()
        val openingSet = setOf('(', '[', '(', '<')
        val map = mapOf(')' to '(', ']' to '[', '>' to '<', ')' to '(')

        for (c in s) {
            // opening parentheses
            if (openingSet.contains(c)) {
                stack.push(c)
            }
            // closing pararentheses
            if (map.containsKey(c)) {
                // De Morgan's laws, complement of (!stack.isEmpty() && map.getValue(c) == stack.peek())
                if (stack.isEmpty() || map.getValue(c) != stack.peek()) {
                    // if more closing parenthese, can already fast-exit
                    return false
                } else {
                    stack.pop()
                }
            }
        }

        return stack.isEmpty()
    }
}
